File: src/system-state-recorder/src/placement_request.rs
#![allow(non_snake_case)]

use serde::Serialize;
#[derive(Serialize)]
struct Capacity {
    CPU: i64,
    DISK_SIZE: i64,
    MEMORY: i64,
}

#[derive(Serialize)]
struct UserTemplate {
    LOGO: String,
    LXD_SECURITY_PRIVILEGED: String,
    SCHED_REQUIREMENTS: String,
}

#[derive(Serialize)]
struct VM {
    CAPACITY: Capacity,
    HOST_IDS: Vec<i32>,
    ID: i64,
    STATE: String,
    USER_TEMPLATE: UserTemplate,
}

#[derive(Serialize)]
struct Data {
    VMS: Vec<VM>,
}

pub fn generate_placement_request_json(
    cpu: i64,
    disk_size: i64,
    memory: i64,
    vmid: i64,
    host_ids: Vec<i32>,
) -> String {
    let requirements: Vec<String> = host_ids.iter().map(|id| format!("ID=\"{}\"", id)).collect();
    let sched_requirements = requirements.join(" | ");

    let data = Data {
        VMS: vec![VM {
            CAPACITY: Capacity {
                CPU: cpu,
                DISK_SIZE: disk_size,
                MEMORY: memory,
            },
            HOST_IDS: host_ids,
            ID: vmid,
            STATE: "PENDING".to_string(),
            USER_TEMPLATE: UserTemplate {
                LOGO: "images/logos/ubuntu.png".to_string(),
                LXD_SECURITY_PRIVILEGED: "true".to_string(),
                SCHED_REQUIREMENTS: sched_requirements,
            },
        }],
    };

    serde_json::to_string_pretty(&data).unwrap()
}

File: EnvServer/pkg/opennebula/placement_request.go
package opennebula

import (
	"encoding/json"
)

type Capacity struct {
	CPU      float64 `json:"CPU"`
	DiskSize int     `json:"DISK_SIZE"`
	Memory   int     `json:"MEMORY"`
}

type UserTemplate struct {
	Logo                  string `json:"LOGO"`
	LxdSecurityPrivileged string `json:"LXD_SECURITY_PRIVILEGED"`
	SchedRequirements     string `json:"SCHED_REQUIREMENTS"`
}

type VM struct {
	Capacity     Capacity     `json:"CAPACITY"`
	HostIDs      []int        `json:"HOST_IDS"`
	ID           int          `json:"ID"`
	State        string       `json:"STATE"`
	UserTemplate UserTemplate `json:"USER_TEMPLATE"`
}

func (vm *VM) ToJSON() (string, error) {
	jsonBytes, err := json.Marshal(vm)
	if err != nil {
		return "", err
	}

	return string(jsonBytes), nil
}

type PlacementRequest struct {
	VMs []VM `json:"VMS"`
}

func (request *PlacementRequest) ToJSON() (string, error) {
	jsonBytes, err := json.Marshal(request)
	if err != nil {
		return "", err
	}

	return string(jsonBytes), nil
}

func (request *PlacementRequest) Equals(request2 *PlacementRequest) bool {
	if request2 == nil {
		return false
	}

	if len(request.VMs) != len(request2.VMs) {
		return false
	}

	for i := 0; i < len(request.VMs); i++ {
		if request.VMs[i].ID != request2.VMs[i].ID &&
			request.VMs[i].State != request2.VMs[i].State &&
			request.VMs[i].Capacity.CPU != request2.VMs[i].Capacity.CPU &&
			request.VMs[i].Capacity.DiskSize != request2.VMs[i].Capacity.DiskSize &&
			request.VMs[i].Capacity.Memory != request2.VMs[i].Capacity.Memory &&
			request.VMs[i].UserTemplate.Logo != request2.VMs[i].UserTemplate.Logo &&
			request.VMs[i].UserTemplate.LxdSecurityPrivileged != request2.VMs[i].UserTemplate.LxdSecurityPrivileged &&
			request.VMs[i].UserTemplate.SchedRequirements != request2.VMs[i].UserTemplate.SchedRequirements &&
			len(request.VMs[i].HostIDs) != len(request2.VMs[i].HostIDs) {
			return false
		}

		for j := 0; j < len(request.VMs[i].HostIDs); j++ {
			if request.VMs[i].HostIDs[j] != request2.VMs[i].HostIDs[j] {
				return false
			}
		}

		return true
	}

	return true
}

func ParsePlacementRequest(jsonString string) (*PlacementRequest, error) {
	var request *PlacementRequest
	//print(jsonString)
	err := json.Unmarshal([]byte(jsonString), &request)
	if err != nil {
		return nil, err
	}

	return request, nil
}
